pkg/graveler/graveler.go

```diff
@@ -1294,11 +1296,141 @@ func (g *KVGraveler) Get(ctx context.Context, repositoryID RepositoryID, ref Ref
 }
 
 func (g *KVGraveler) Set(ctx context.Context, repositoryID RepositoryID, branchID BranchID, key Key, value Value, writeConditions ...WriteConditionOption) error {
-	return g.db.Set(ctx, repositoryID, branchID, key, value, writeConditions...)
+	isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repositoryID, branchID, BranchProtectionBlockedAction_STAGING_WRITE)
+	if err != nil {
+		return err
+	}
+	if isProtected {
+		return ErrWriteToProtectedBranch
+	}
+
+	setFunc := func(branch *Branch) error {
+		writeCondition := &WriteCondition{}
+		for _, cond := range writeConditions {
+			cond(writeCondition)
+		}
+
+		if !writeCondition.IfAbsent {
+			return g.StagingManager.Set(ctx, branch.StagingToken, key, &value, !writeCondition.IfAbsent)
+		}
+
+		// check if the given key exist in the branch first
+		_, err := g.Get(ctx, repositoryID, Ref(branchID), key)
+		if err == nil {
+			// we got a key here already!
+			return ErrPreconditionFailed
+		}
+		if !errors.Is(err, ErrNotFound) {
+			// another error occurred!
+			return err
+		}
+
+		return g.StagingManager.Update(ctx, branch.StagingToken, key, func(v *Value) (*Value, error) {
+			if v == nil || v.Identity == nil {
+				// value doesn't exist or is a tombstone
+				return &value, nil
+			}
+			return nil, ErrPreconditionFailed
+		})
+	}
+
+	return g.safeBranchWrite(ctx, g.log.WithField("key", key).WithField("operation", "set"), repositoryID, branchID, setFunc)
+}
+
+// safeBranchWrite is a helper function that wraps a branch write operation with validation that the staging token
+// didn't change while writing to the branch.
+func (g *KVGraveler) safeBranchWrite(ctx context.Context, log logging.Logger, repositoryID RepositoryID, branchID BranchID, stagingOperation func(branch *Branch) error) error {
+	// setTries is the number of times to repeat the set operation if the staging token changed
+	const setTries = 3
```

Comment:
Maybe we should export the retry mechanism?

Reply:
The only other usage I see for it is the `Commit` and it can use it from here too, no?